File: problems/1-easy/543-Diameter-Of-Binary-Tree/solution.cpp
#include <algorithm>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

int max_diameter = 0;
int diameterOfBinaryTree(TreeNode* root) {
    max_diameter = 0;
    dfs(root);
    return max_diameter;
}

int dfs(TreeNode* root) {
    if (!root) {
        return -1;
    }
    
    int max_left = dfs(root->left);
    int max_right = dfs(root->right);

    max_diameter = max(max_diameter, max_left + max_right + 2);
    return max(max_left, max_right) + 1;

}

File: problems/1-easy/543-Diameter-Of-Binary-Tree/solution.py
from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:

        self.res = 0
        
        self.maxDiameter(root)

        return self.res
        
    def maxDiameter(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0
        
        left_depth = self.maxDiameter(root.left)

        right_depth = self.maxDiameter(root.right)

        self.res = max(left_depth + right_depth, self.res)

        return max(left_depth,right_depth) + 1